Applications/spire/source/charting/cached_chart_model.cpp

```diff
@@ -0,0 +1,125 @@
+}
+
+QtPromise<std::vector<Candlestick>> CachedChartModel::load_data(
+    const ChartRange& data) {
+  return m_chart_model->load(data.m_start, data.m_end).then(
+    [=] (auto result) {
+      auto first = std::lower_bound(m_loaded_data.begin(),
+        m_loaded_data.end(), data.m_start, [] (const auto& index,
+            const auto& value) {
+          return index.GetStart() < value;
+        });
+      auto last = std::lower_bound(m_loaded_data.begin(),
+        m_loaded_data.end(), data.m_end, [] (const auto& index,
+            const auto& value) {
+          return index.GetStart() < value;
+        });
+      auto pos = m_loaded_data.end();
+      if(last != m_loaded_data.end()) {
+        pos = m_loaded_data.erase(first, last + 1);
+      } else {
+        pos = m_loaded_data.erase(first, last);
+      }
+      m_loaded_data.insert(pos, result.Get().begin(), result.Get().end());
+      auto [range_first, range_last] = range_search(data);
+      auto new_range = data;
+      if(range_first != m_ranges.end()) {
+        new_range.m_start = std::min((*range_first).m_start, new_range.m_start);
+      }
+      if(range_last != m_ranges.end()) {
+        new_range.m_end = std::max((*range_last).m_end, new_range.m_end);
+      }
+      auto range_pos = m_ranges.erase(range_first, range_last);
+      m_ranges.insert(range_pos, new_range);
+      return result.Get();
+  });
+}
+
+std::tuple<std::vector<CachedChartModel::ChartRange>::iterator,
+    std::vector<CachedChartModel::ChartRange>::iterator>
+    CachedChartModel::range_search(const ChartRange& range) {
+  if(m_ranges.empty()) {
+    return {m_ranges.end(), m_ranges.end()};
+  }
+  auto first = std::optional<std::vector<ChartRange>::iterator>();
+  auto last = std::optional<std::vector<ChartRange>::iterator>();
+  if(range.m_start < m_ranges.front().m_start) {
+    first = m_ranges.begin();
+    if(range.m_end < m_ranges.front().m_start) {
+      return {first.value(), first.value()};
+    }
+  }
+  if(range.m_end > m_ranges.back().m_end) {
+    last = m_ranges.end();
+    if(range.m_start > m_ranges.back().m_end) {
+      return {last.value(), last.value()};
+    }
+  }
+  for(auto loaded_range = m_ranges.begin(); loaded_range != m_ranges.end();
+      ++loaded_range) {
+    if((*loaded_range).m_start <= range.m_start &&
+        (*loaded_range).m_end >= range.m_start) {
+      first = loaded_range;
+    }
+    if((*loaded_range).m_start <= range.m_end &&
+        (*loaded_range).m_end >= range.m_end) {
+      last = loaded_range + 1;
+    }
+    if(!first.has_value() && range.m_start > (*loaded_range).m_end &&
+        range.m_start < (*(loaded_range + 1)).m_start) {
+      first = loaded_range + 1;
+    }
+    if(!last.has_value() && range.m_end > (*loaded_range).m_end &&
+        range.m_end < (*(loaded_range + 1)).m_start) {
+      last = loaded_range + 1;
+    }
+    if(first.has_value() && last.has_value()) {
+      break;
+    }
+  }
+  return {first.value(), last.value()};
```

Comment:
How do you know ```first``` and ```last``` have a ```value()```?

Reply:
Didn't we exhaustively prove that by the time we've reached line 124, first and last have to have values?

Reply:
No and if you did, feel free to show it.

Reply:
The cases we had were:

1. The requested range, P, is a subset of a range in the previously loaded ranges, R. This can't be the case in load_data() because it's handled before load_data() can be called.

2. R is empty, so (R.end, R.end) is returned.

3. P is to the left of R[0].

4. P is to the right of R.end().

5. P.start is to the left of R[0].

6. P.end is to the right of R.end().

7. P.start and P.end both fall within a range in R.

8. P.start is within a range in R, and P.end is in between two ranges.

9. P.start is between two ranges, and P.end is within a range in R.

10. Both P.start and P.end are between a range in R.

As far as I can tell this covers every possible case, so at line 124 first and last can't be null.

Reply:
If ```P``` is to the right of ```R.end()```, then what are ```first``` and ```last``` initialized to?

Reply:
That case is evaluated on lines 96 and 98. In that case last = m_ranges.end(), and (last, last) is returned.